### ISSUE TITLE
ocularAutoThreshold

### ISSUE THREAD

**Issue by hafedh-trimeche** [warrengalyen/ocular#1]
Hello,

Would a demo for **ocularAutoThreshold** be provided?

Best regards.

**Comment by warrengalyen** [comment]
I've just added an example to the docs.

https://www.mechanikadesign.com/docs/ocular/1.0/page_examples_autothreshold.html

**Comment by hafedh-trimeche** [comment]
Hi,

**ocularAutoThreshold** solved.

Using this Delphi code
```
function ocularAutoContrast(Bytes:TBytes):TBytes;
var
  Input    : Pointer;
  Bmp      : TBitmap;
  Width    ,
  Height   ,
  Channels : Integer;
begin
  if Bytes=nil then Exit(nil);
  InitUnit;
  Input := stbi_load_from_memory(@Bytes[0],Length(Bytes),@Width,@Height,@Channels,0);
  if Input=nil then Exit(nil);
  SetLength(Result,Width*Height*Channels);
  _ocularAutoContrast(Input,Result,Width,Height,Width*Channels);
  stbi_image_free(Input);
  Bmp := TBitmap.Create;
  Bmp.Load(Result,Width,Height,Channels*8);
  Result := Bmp.Save;
  FreeAndNil(Bmp);
end;

function ocularAutoGammaCorrection(Bytes:TBytes):TBytes;
var
  Input    : Pointer;
  Bmp      : TBitmap;
  Width    ,
  Height   ,
  Channels : Integer;
begin
  if Bytes=nil then Exit(nil);
  InitUnit;
  Input := stbi_load_from_memory(@Bytes[0],Length(Bytes),@Width,@Height,@Channels,0);
  if Input=nil then Exit(nil);
  SetLength(Result,Width*Height*Channels);
  _ocularAutoGammaCorrection(Input,Result,Width,Height,Width*Channels);
  stbi_image_free(Input);
  Bmp := TBitmap.Create;
  Bmp.Load(Result,Width,Height,Channels*8);
  Result := Bmp.Save;
  FreeAndNil(Bmp);
end;

function ocularAutoThreshold(Bytes:TBytes;Method:OcAutoThresholdMethod):TBytes;
var
  Input    : Pointer;
  Bmp      : TBitmap;
  Width    ,
  Height   ,
  Channels : Integer;
begin
  if Bytes=nil then Exit(nil);
  InitUnit;
  Input := stbi_load_from_memory(@Bytes[0],Length(Bytes),@Width,@Height,@Channels,0);
  if Input=nil then Exit(nil);
  _ocularGrayscaleFilter(Input,Input,Width,Height,Width*Channels);
  Channels := 1;
  SetLength(Result,Width*Height*Channels);
  _ocularAutoThreshold(Input,Result,Width,Height,Width,Method);
  stbi_image_free(Input);
  Bmp := TBitmap.Create;
  Bmp.Load(Result,Width,Height,Channels*8);
  Result := Bmp.Save;
  FreeAndNil(Bmp);
end;
```
only **ocularAutoContrast** raised an exception.

Best regards.

**Comment by warrengalyen** [comment]
I see your problem. The `ocularAutoContrast` function accepts channels as the last parameter. Remove the width since the stride is being calculated inside the function. Sorry for the confusion. When migrating some of my code I used channels instead of the normal stride parameter.  I'll make this more uniform at some point. Refer to the filter function definitions in the documentation.

**Comment by hafedh-trimeche** [comment]
Thanks.

Would auto image enhancement be considered (OCR, Fingerprint Recognition, Face Recognition, ...)?

**Comment by warrengalyen** [comment]
Auto enhancement would be considered color adjustment image processing filters.

**Comment by hafedh-trimeche** [comment]
Please find OCR preprocess flow:
https://nextgeninvent.com/blogs/7-steps-of-image-pre-processing-to-improve-ocr-using-python-2/

**Comment by warrengalyen** [comment]
Thanks for the reference but I've already started researching some more advanced methods for OCR pre-processing. I have a list of features I wanted to incorporate into the library but I have will to develop most of the algorithms from scratch using research papers so it will take quite a bit of time. The main focus of the library is processing photos, so I don't want to go too off track with OCR.

I've already incorporated deskewing and despeckle (just added) but these are some possible features:

- Detect and remove black (dark) edges from a scanned image. Border removal
- Remove stamps/seals
- Remove hole punches (left, right, top)
- Blank page detection
- Line removal, overlap on text
